Index: hpnssh-8.8p1_hpn16v1/channels.c
===================================================================
--- hpnssh-8.8p1_hpn16v1.orig/channels.c
+++ hpnssh-8.8p1_hpn16v1/channels.c
@@ -4657,6 +4657,16 @@ x11_create_display_inet(struct ssh *ssh,
 				debug2_f("bind port %d: %.100s", port,
 				    strerror(errno));
 				close(sock);
+
+				/* do not remove successfully opened
+				 * sockets if the request failed because
+				 * the protocol IPv4/6 is not available
+				 * (e.g. IPv6 may be disabled while being
+				 * supported)
+				 */
+				if (EADDRNOTAVAIL == errno)
+    					continue;
+
 				for (n = 0; n < num_socks; n++)
 					close(socks[n]);
 				num_socks = 0;
