Index: hpnssh-8.8p1_hpn16v1/scp.c
===================================================================
--- hpnssh-8.8p1_hpn16v1.orig/scp.c
+++ hpnssh-8.8p1_hpn16v1/scp.c
@@ -1166,7 +1166,9 @@ toremote(int argc, char **argv, enum scp
 			addargs(&alist, "%s", host);
 			addargs(&alist, "%s", cmd);
 			addargs(&alist, "%s", src);
-			addargs(&alist, "%s%s%s:%s",
+			addargs(&alist,
+			    /* IPv6 address needs to be enclosed with sqare brackets */
+			    strchr(host, ':') != NULL ? "%s%s[%s]:%s" : "%s%s%s:%s",
 			    tuser ? tuser : "", tuser ? "@" : "",
 			    thost, targ);
 			if (do_local_cmd(&alist) != 0)
